manifests/value.pp

```diff
@@ -57,7 +57,8 @@
   validate_re($type, '^\w+',
     'type parameter must not be empty but it is type => "$type"')
 
-
+  # normalize the case so that our defined() check is more consistent
+  $key_normalized = downcase($key)
```

Comment:
It's an interesting proposition - comparisons are insensitive, but setting values is case preserving.

If the key matches case insensitively, then it implies values should not be modified by Puppet.  But I agree with @ferventcoder that when the value *is* being set, it should apply the case that the user set in their manifest.